core/src/main/java/org/apache/carbondata/core/indexstore/blockletindex/BlockletDataRefNode.java

```diff
@@ -133,35 +139,107 @@ public short blockletIndex() {
+    return chunk;
+  }
+
+  private DimensionRawColumnChunk readDimensionChunkWithoutCache(
+      FileReader fileReader, int columnIndex) throws IOException {
+    DimensionColumnChunkReader reader = getDimensionColumnChunkReader(fileReader);
+    return reader.readRawDimensionChunk(fileReader, columnIndex);
+  }
+
+  private MeasureRawColumnChunk readMeasureChunkWithoutCache(
+      FileReader fileReader, int columnIndex) throws IOException {
+    MeasureColumnChunkReader measureColumnChunkReader = getMeasureColumnChunkReader(fileReader);
+    return measureColumnChunkReader.readRawMeasureChunk(fileReader, columnIndex);
   }
 
   @Override
-  public MeasureRawColumnChunk[] readMeasureChunks(FileReader fileReader, int[][] columnIndexRange)
+  public MeasureRawColumnChunk[] readMeasureChunks(FileReader fileReader, int[][] columnIndexes)
       throws IOException {
-    MeasureColumnChunkReader measureColumnChunkReader = getMeasureColumnChunkReader(fileReader);
-    MeasureRawColumnChunk[] measureRawColumnChunks =
-        measureColumnChunkReader.readRawMeasureChunks(fileReader, columnIndexRange);
-    return measureRawColumnChunks;
+    TableBlockInfo blockInfo = blockInfos.get(index);
+    if (ColumnChunkCache.isEnabledForTable(tableId)) {
+      MeasureRawColumnChunk[] chunks = new MeasureRawColumnChunk[
+          blockInfo.getDetailInfo().getBlockletInfo().getMeasureChunkOffsets().size()];
+      for (int[] columnIndex : columnIndexes) {
+        int columnIndexStart = columnIndex[0];
+        int columnIndexEnd = columnIndex[1];
+        for (int j = columnIndexStart; j <= columnIndexEnd; j++) {
+          ColumnChunkCache.CacheKey key = new ColumnChunkCache.CacheKey(
+              blockInfo.getFilePath(),
+              blockInfo.getDetailInfo().getBlockletInfo().getMeasureChunkOffsets().get(j));
+          Optional<AbstractRawColumnChunk> chunkOp = ColumnChunkCache.get(tableId, key);
+          if (chunkOp.isPresent()) {
+            chunks[j] = (MeasureRawColumnChunk) chunkOp.get();
+          } else {
+            chunks[j] = readMeasureChunkWithoutCache(fileReader, j);
+            ColumnChunkCache.put(tableId, key, chunks[j]);
+          }
+        }
+      }
+      return chunks;
+    } else {
+      MeasureColumnChunkReader measureColumnChunkReader = getMeasureColumnChunkReader(fileReader);
+      return measureColumnChunkReader.readRawMeasureChunks(fileReader, columnIndexes);
+    }
   }
 
   @Override
   public MeasureRawColumnChunk readMeasureChunk(FileReader fileReader, int columnIndex)
       throws IOException {
-    MeasureColumnChunkReader measureColumnChunkReader = getMeasureColumnChunkReader(fileReader);
-    MeasureRawColumnChunk measureRawColumnChunk =
-        measureColumnChunkReader.readRawMeasureChunk(fileReader, columnIndex);
-    return measureRawColumnChunk;
+    TableBlockInfo blockInfo = blockInfos.get(index);
+    ColumnChunkCache.CacheKey key = new ColumnChunkCache.CacheKey(
+        blockInfo.getFilePath(),
```

Comment:
If the task distribution is '`blocklet`' instead of `block` to get the splits in the driver. Two splits may have  same block but different blocklet id. In that cache will be overwritten for the last blokclet.

Reply:
I mean we cannot use blockInfo.getFilePath() as a cache key in case of `bocklet` distribution